Fix layering violation in Frame::disconnectOwnerElement()

charliewolfe opened the pull request:
#### d20a6a2130032e592e10109d6a4617f7b741e684
<pre>
Fix layering violation in Frame::disconnectOwnerElement()
<a href="https://bugs.webkit.org/show_bug.cgi?id=263948">https://bugs.webkit.org/show_bug.cgi?id=263948</a>
<a href="https://rdar.apple.com/117721204">rdar://117721204</a>

Reviewed by Pascoe.

Document should only be used from LocalFrame.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::frameWasDisconnectedFromOwner): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::disconnectOwnerElement):
* Source/WebCore/page/Frame.h:
(WebCore::Frame::frameWasDisconnectedFromOwner const):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::frameWasDisconnectedFromOwner const):
* Source/WebCore/page/LocalFrame.h:

Canonical link: <a href="https://commits.webkit.org/270042@main">https://commits.webkit.org/270042@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59ace1ee57e6c8214c85688272a1f58d1afab083

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24329 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25416 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26462 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22394 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24598 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4078 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22829 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1960 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27052 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1704 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21950 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28162 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22185 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22259 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25958 "Found 2 new API test failures: /WebKitGTK/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WebKitGTK/TestLoaderClient:/webkit/WebKitURIRequest/http-headers (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1629 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2 "Found 1 new test failure: fast/replaced/pdf-as-embed-with-no-mime-type-is-not-blank.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/2908 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2035 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3111 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1997 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1231,6 +1231,17 @@ DataDetectionResultsStorage& LocalFrame::dataDetectionResults()
+    if (RefPtr window = m_doc->domWindow())
+        window->willDetachDocumentFromFrame();
+
+    m_doc->detachFromFrame();
```
Can we protect the document before calling `detachFromFrame()` on it. This doesn't match our modern smart pointer idioms.

cdumez replied:
`protectedDocument()->detachFromFrame();` (unless it is possible for the document to be in the middle of destruction).

rniwa commented on Source/WebCore/page/Frame.h:
```diff
@@ -89,6 +89,8 @@ class Frame : public ThreadSafeRefCounted<Frame, WTF::DestructionThread::Main>,
     Frame(Page&, FrameIdentifier, FrameType, HTMLFrameOwnerElement*, Frame* parent);
     void resetWindowProxy();
 
+    virtual void frameWasDisconnectedFromOwner() const { };
```
Nit: no need for ;

cdumez commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1232,6 +1232,17 @@ DataDetectionResultsStorage& LocalFrame::dataDetectionResults()
+    if (RefPtr window = m_doc->domWindow())
+        window->willDetachDocumentFromFrame();
+
+    protectedDocument()->detachFromFrame();
```
Please make sure to wait for debug EWS before landing. It is often not easy to tell if a particular code path can get called by the Document destructor.

charliewolfe committed:
Fix layering violation in Frame::disconnectOwnerElement()
https://bugs.webkit.org/show_bug.cgi?id=263948
rdar://117721204

Reviewed by Pascoe.

Document should only be used from LocalFrame.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::frameWasDisconnectedFromOwner): Deleted.
* Source/WebCore/dom/Document.h:
* Source/WebCore/page/Frame.cpp:
(WebCore::Frame::disconnectOwnerElement):
* Source/WebCore/page/Frame.h:
(WebCore::Frame::frameWasDisconnectedFromOwner const):
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::frameWasDisconnectedFromOwner const):
* Source/WebCore/page/LocalFrame.h:

Canonical link: https://commits.webkit.org/270042@main
